sympy/codegen/ast.py

```diff
@@ -0,0 +1,732 @@
+            if lhs in left_hand_sides:
+                raise NotImplementedError("Duplicate assignments to the same "
+                "variable are not yet supported (%s)" % lhs)
+            left_hand_sides.append(lhs)
+            right_hand_sides.append(rhs)
+
+        obj = Basic.__new__(cls, *assignments)
+
+        obj.left_hand_sides = Tuple(*left_hand_sides)
+        obj.right_hand_sides = Tuple(*right_hand_sides)
+
+        return obj
+
+    @classmethod
+    def topological_sort(cls, assignments):
+        """
+        Return a CodeBlock with topologically sorted assignments so that
+        variables are assigned before they are used.
+
+        The existing order of assignments is preserved as much as possible.
+
+        This is a class constructor so that the default constructor for
+        CodeBlock can error when variables are used before they are assigned.
+
+        Example
+        =======
+
+        >>> from sympy import symbols
+        >>> from sympy.codegen.ast import CodeBlock, Assignment
+        >>> x, y, z = symbols('x y z')
+
+        >>> assignments = [
+        ...     Assignment(x, y + z),
+        ...     Assignment(y, z + 1),
+        ...     Assignment(z, 2),
+        ... ]
+        >>> CodeBlock.topological_sort(assignments)
+        CodeBlock(Assignment(z, 2), Assignment(y, z + 1), Assignment(x, y + z))
+        """
+        from sympy.utilities.iterables import topological_sort
+        # Create a graph where the nodes are assignments and there is a directed edge
+        # between nodes that use a variable and nodes that assign that
+        # variable, like
+
+        # [(x := 1, y := x + 1), (x := 1, z := y + z), (y := x + 1, z := y + z)]
+
+        # If we then topologically sort these nodes, they will be in
+        # assignment order, like
+
+        # x := 1
+        # y := x + 1
+        # z := y + z
+
+        # A = The nodes
+        #
+        # enumerate keeps nodes in the same order they are already in if
+        # possible. It will also allow us to handle duplicate assignments to
+        # the same variable when those are implemented.
+        A = list(enumerate(assignments))
+
+        # var_map = {variable: [assignments using variable]}
+        # like {x: [y := x + 1, z := y + x], ...}
+        var_map = {}
+
+        # E = Edges in the graph
+        E = []
+        for i in A:
+            if i[1].lhs in var_map:
+                E.append((var_map[i[1].lhs], i))
+            var_map[i[1].lhs] = i
+        for i in A:
+            for x in i[1].rhs.free_symbols:
+                if x not in var_map:
+                    # XXX: Allow this case?
+                    raise ValueError("Undefined variable %s" % x)
+                E.append((var_map[x], i))
+
+        ordered_assignments = topological_sort([A, E])
+        # De-enumerate the result
+        return cls(*list(zip(*ordered_assignments))[1])
+
+    def cse(self, symbols=None, optimizations=None, postprocess=None,
```

Comment:
For this to be useful in a C setting it needs to somehow return the new variables to the user so that they can be initialized properly.